[1205] FhirEngine Sync Download API Re-design

omarismail94 opened the pull request:
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1205 

**Description**
FhirEngine Sync Download API Redesign

See for details: https://docs.google.com/document/d/1UZ7Ndb8JRmB2UdXpfIkPT-VAra6X2aLSzdGI_a8KJkg/edit#

TESTED=all unit tests passed. Added one more unit test as well

tl;dr: give more functionality to developers to control what resources they can download.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?
No, open to suggestions though!

**Type**
Choose one: (Bug fix | **Feature** | Documentation | Testing | Code health | Builds | Releases | Other)


**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

aditya-07 commented on engine/src/main/java/com/google/android/fhir/sync/FhirSyncWorker.kt:
```diff
@@ -35,22 +35,34 @@ import kotlinx.coroutines.cancel
   abstract fun getFhirEngine(): FhirEngine
   abstract fun getDataSource(): DataSource
-  abstract fun getSyncData(): ResourceSyncParams
+  abstract fun getInitialUrl(): String
+  abstract fun getCreateDownloadUrl(): (String, String?) -> String
+  abstract fun getExtractResourcesFromResponse(): (Resource) -> Collection<Resource>
+  abstract fun getExtractNextUrlsFromResource(): (Resource) -> Collection<String>
```
Should we encapsulate this in something like a `SyncDownloadContext`? Engine can provide a default implementation for it based on what we have implemented in `FhirPeriodicSyncWorker` and Developer  applications  may use it or provide their custom implementation based on the use case?

omarismail94 replied:
Great idea! I created an interface named `SyncDownloadExtractor` that users can implement. Provided a sample implementation in the `demo` code, This interface is need to instantiate a `FhirSyncWorker`

Tarun-Bhardwaj commented:
@omarismail94 , did you get a chance to work on the review comments? This issue is lined up for FHIR Engine Beta release next week.
CC: @jingtang10

omarismail94 commented:
@aditya-07 lmk if there are other comments that need addressing. I created a new interface named `SyncDownloadExtractor` that users can implement. An implementation of this interface is needed to instantiate a `FhirSyncWorker`. 

Note that I refactored the functions to take in parameters now as opposed to creating lambdas to make things cleaner


> @omarismail94 , did you get a chance to work on the review comments? This issue is lined up for FHIR Engine Beta release next week. CC: @jingtang10

@Tarun-Bhardwaj  yes :)

jingtang10 commented on demo/src/main/java/com/google/android/fhir/demo/data/SyncDownloadExtractorImpl.kt:
```diff
@@ -0,0 +1,86 @@
+  override fun createDownloadUrl(preProcessUrl: String, lastUpdate: String?): String {
+    var downloadUrl = preProcessUrl
+
+    if (lastUpdate != null && downloadUrl.contains("\$everything")) {
+      downloadUrl = "$downloadUrl?_since=$lastUpdate"
+    }
+
+    if (lastUpdate != null && !downloadUrl.contains("\$everything")) {
+      downloadUrl = "$downloadUrl&_lastUpdated=gt$lastUpdate"
+    }
+
+    if (downloadUrl.contains("&page_token")) {
+      downloadUrl = preProcessUrl
```
some documentation / comments would be great for this

jingtang10 commented on demo/src/main/java/com/google/android/fhir/demo/data/SyncDownloadExtractorImpl.kt:
```diff
@@ -0,0 +1,86 @@
+  override fun extractNextUrlsFromResource(resourceResponse: Resource): Collection<String> {
+    val queueWork = mutableListOf<String>()
+
+    if (resourceResponse is ListResource) {
+      for (entry in resourceResponse.entry) {
+        val patientUrl = "${entry.item.reference}/\$everything"
+        queueWork.add(patientUrl)
+      }
+    }
+
+    if (resourceResponse is Bundle) {
+      val nextUrl =
+        resourceResponse.link.firstOrNull { component -> component.relation == "next" }?.url
+      if (nextUrl != null) {
+        queueWork.add(nextUrl)
+      }
+    }
```
add a brief comment to explain the 2 cases

jingtang10 commented on engine/src/test/java/com/google/android/fhir/sync/SyncJobTest.kt:
```diff
@@ -146,7 +148,7 @@ class SyncJobTest {
     val job = launch { flow.collect { res.add(it) } }
 
-    syncJob.run(fhirEngine, dataSource, resourceSyncParam, flow)
+    syncJob.run(fhirEngine, dataSource, TestingUtils.TestSyncDownloadExtractorImpl, flow)
```
the `resourceSyncParam` in this class can be deleted?

jingtang10 reviewed (changes requested):
are there no tests to collect the flow of the resources? do you think we should add a test class for fhirsynchronizer?

jingtang10 commented on engine/src/main/java/com/google/android/fhir/sync/SyncDownloadExtractor.kt:
```diff
@@ -16,17 +16,12 @@
 
-class TestFailingSyncWorker(appContext: Context, workerParams: WorkerParameters) :
-  FhirSyncWorker(appContext, workerParams) {
+interface SyncDownloadExtractor {
```
just brainstorming, this is optional :)
```suggestion
interface DownloadProcessor {
```


```suggestion
interface DownloadHandler {
```

omarismail94 replied:
I saw the `FhirSynchronizer` contains a reference to `Uploader`, so I changed the name to `Downloader`. Lmk if that works!

jingtang10 replied:
i think the uploader does something different, it actually does the uploading!

i think here maybe download processor is better. @aditya-07 even suggested DownloadPrePostProcessor, or DownloadManager?

omarismail94 replied:
Renamed to `DownloadManager`!

aditya-07 commented on engine/src/main/java/com/google/android/fhir/sync/FhirSynchronizer.kt:
```diff
@@ -102,14 +103,35 @@ internal class FhirSynchronizer(
+        listOfUrls.add(syncDownloadExtractor.getInitialUrl())
+        while (listOfUrls.isNotEmpty()) {
+          try {
+            val preprocessedUrl = listOfUrls.removeAt(0)
```
Better make `listOfUrls` a `Queue<String>` and use appropriate apis.

omarismail94 replied:
Changed to a queue. Good idea!

omarismail94 commented:
> are there no tests to collect the flow of the resources? do you think we should add a test class for fhirsynchronizer?

 @jingtang10  Added a test to count the number of times resources get loaded that should be equal to the length of the queue

Also addressed all comments you made. Lmk if there's anything else you want!

jingtang10 reviewed (approved):
thanks @omarismail94!
